raft/raft_test.go

```diff
@@ -4141,6 +4141,218 @@ func TestPreVoteMigrationWithFreeStuckPreCandidate(t *testing.T) {
 	}
 }
 
+func TestFastLogRejection(t *testing.T) {
+	tests := []struct {
+		logs  []pb.Entry
+		ent1  []pb.Entry
+		ent2  []pb.Entry
+		term  uint64
+		index uint64
+	}{
+		// append reject, return index that is not conflict
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 4, Index: 4},
+				{Term: 4, Index: 5},
+			},
+			[]pb.Entry{
+				{Term: 4, Index: 6},
+				{Term: 4, Index: 7},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 3, Index: 4},
+				{Term: 3, Index: 5},
+				{Term: 3, Index: 6},
+				{Term: 3, Index: 7},
+				{Term: 3, Index: 8},
+				{Term: 3, Index: 9},
+				{Term: 3, Index: 10},
+				{Term: 3, Index: 11},
+			},
+			2,
+			3,
+		},
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 3, Index: 4},
+				{Term: 4, Index: 5},
+				{Term: 4, Index: 6},
+			},
+			[]pb.Entry{
+				{Term: 4, Index: 7},
+				{Term: 5, Index: 8},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 3, Index: 4},
+				{Term: 3, Index: 5},
+				{Term: 3, Index: 6},
+				{Term: 3, Index: 7},
+				{Term: 3, Index: 8},
+				{Term: 3, Index: 9},
+				{Term: 3, Index: 10},
+				{Term: 3, Index: 11},
+			},
+			3,
+			4,
+		},
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 3, Index: 3},
+			},
+			[]pb.Entry{
+				{Term: 4, Index: 4},
+				{Term: 5, Index: 5},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+				{Term: 4, Index: 4},
+			},
+			1,
+			2,
+		},
+		// new leader index is smaller than old leader at index 2
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 3, Index: 3},
+			},
+			[]pb.Entry{
+				{Term: 4, Index: 4},
+				{Term: 4, Index: 5},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 2, Index: 2},
+				{Term: 2, Index: 3},
+				{Term: 4, Index: 4},
+			},
+			1,
+			2,
+		},
+		// append successfully
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+			},
+			[]pb.Entry{
+				{Term: 4, Index: 4},
+				{Term: 5, Index: 5},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+				{Term: 4, Index: 4},
+			},
+			5,
+			5,
+		},
+		{
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+			},
+			[]pb.Entry{
+				{Term: 4, Index: 4},
+				{Term: 5, Index: 5},
+			},
+			[]pb.Entry{
+				{Term: 1, Index: 1},
+				{Term: 1, Index: 2},
+				{Term: 1, Index: 3},
+				{Term: 3, Index: 4},
+			},
+			5,
+			5,
+		},
+	}
+
+	for i, test := range tests {
+		s1 := NewMemoryStorage()
+		s1.Append(test.logs)
+
+		n1 := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, s1)
+		n2 := newTestRaft(2, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())
```

Comment:
are you intentionally not appending `test.logs` on n2? I thought test.logs was a shared prefix and ents1 and ents2 would be possibly different suffixes. It doesn't seem to matter for this test but seems cleaner (less confusing) anyway.